File: packages/contracts-ts/src/features.ts
import { z } from 'zod';

import { authoredIdSchema, eventIdSchema, runIdSchema, sequenceSchema } from './primitives';
import {
  DATASET_MANIFEST_SCHEMA_VERSION,
  FEATURE_SCHEMA_MANIFEST_SCHEMA_VERSION,
  FEATURE_VECTOR_SCHEMA_VERSION,
  FEATURE_WINDOW_SCHEMA_VERSION,
  ONLINE_FEATURE_UPDATE_SCHEMA_VERSION,
} from './versioning';

export const TRANSFORM_VERSION = '0.0.0-phase14' as const;
export const DEFAULT_WINDOW_DURATION_SIM_SECONDS = 300;

export const FeatureErrorCode = {
  VALIDATION_FAILED: 'FEATURE_VALIDATION_FAILED',
  UNSUPPORTED_EVENT: 'FEATURE_UNSUPPORTED_EVENT',
  DUPLICATE_EVENT: 'FEATURE_DUPLICATE_EVENT',
  OUT_OF_ORDER: 'FEATURE_OUT_OF_ORDER',
  LATE_EVENT: 'FEATURE_LATE_EVENT',
  HIDDEN_TRUTH_BLOCKED: 'FEATURE_HIDDEN_TRUTH_BLOCKED',
  SCHEMA_MISMATCH: 'FEATURE_SCHEMA_MISMATCH',
  MALFORMED_PAYLOAD: 'FEATURE_MALFORMED_PAYLOAD',
} as const;

export const featureErrorCodeSchema = z.enum([
  'FEATURE_VALIDATION_FAILED',
  'FEATURE_UNSUPPORTED_EVENT',
  'FEATURE_DUPLICATE_EVENT',
  'FEATURE_OUT_OF_ORDER',
  'FEATURE_LATE_EVENT',
  'FEATURE_HIDDEN_TRUTH_BLOCKED',
  'FEATURE_SCHEMA_MISMATCH',
  'FEATURE_MALFORMED_PAYLOAD',
]);

export const featureDataTypeSchema = z.enum(['float', 'int', 'categorical']);

export const featureDefinitionSchema = z
  .object({
    name: z.string().min(1).max(128),
    orderIndex: z.number().int().min(0),
    dtype: featureDataTypeSchema,
    unit: z.string().min(1).max(64),
    minValue: z.number().nullable().optional(),
    maxValue: z.number().nullable().optional(),
    missingSentinel: z.number(),
    categoricalMappingVersion: z.number().int().min(1).nullable().optional(),
    categoricalValues: z.array(z.string()).nullable().optional(),
  })
  .strict();

export const featureSchemaManifestSchema = z
  .object({
    schemaVersion: z.number().int().min(1),
    featureSchemaVersion: z.number().int().min(1),
    transformVersion: z.string().min(1),
    windowDurationSimSeconds: z.number().int().min(1),
    features: z.array(featureDefinitionSchema).min(1),
  })
  .strict()
  .superRefine((value, ctx) => {
    if (value.schemaVersion !== FEATURE_SCHEMA_MANIFEST_SCHEMA_VERSION) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message: `Unsupported feature schema manifest version: ${String(value.schemaVersion)}`,
      });
    }
    const indices = value.features.map((feature) => feature.orderIndex);
    if (indices.join(',') !== [...indices].sort((a, b) => a - b).join(',')) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message: 'Feature definitions must be sorted by orderIndex',
      });
    }
    if (new Set(indices).size !== indices.length) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message: 'Feature orderIndex values must be unique',
      });
    }
  });

export const featureProvenanceSchema = z
  .object({
    runId: runIdSchema,
    entityId: authoredIdSchema,
    sequenceStart: sequenceSchema,
    sequenceEnd: sequenceSchema,
    simTimeStart: z.string().min(1),
    simTimeEnd: z.string().min(1),
    featureSchemaVersion: z.number().int().min(1),
    transformVersion: z.string().min(1),
    sourceEventIds: z.array(eventIdSchema),
  })
  .strict();

export const featureWindowSchema = z
  .object({
    schemaVersion: z.number().int().min(1),
    windowKey: z.string().min(1),
    runId: runIdSchema,
    entityId: authoredIdSchema,
    windowStartSimTime: z.string().min(1),
    windowEndSimTime: z.string().min(1),
    isClosed: z.boolean(),
    watermarkSequence: sequenceSchema,
  })
  .strict()
  .superRefine((value, ctx) => {
    if (value.schemaVersion !== FEATURE_WINDOW_SCHEMA_VERSION) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message: `Unsupported feature window schema version: ${String(value.schemaVersion)}`,
      });
    }
  });

export const featureVectorSchema = z
  .object({
    schemaVersion: z.number().int().min(1),
    featureSchemaVersion: z.number().int().min(1),
    windowKey: z.string().min(1),
    entityId: authoredIdSchema,
    values: z.array(z.number()),
    provenance: featureProvenanceSchema,
  })
  .strict()
  .superRefine((value, ctx) => {
    if (value.schemaVersion !== FEATURE_VECTOR_SCHEMA_VERSION) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message: `Unsupported feature vector schema version: ${String(value.schemaVersion)}`,
      });
    }
  });

export const datasetSourceRunSchema = z
  .object({
    runId: runIdSchema,
    seed: z.number().int(),
    scenarioVersionId: authoredIdSchema,
  })
  .strict();

export const datasetManifestSchema = z
  .object({
    schemaVersion: z.number().int().min(1),
    datasetChecksum: z.string().regex(/^sha256:[0-9a-f]{64}$/),
    featureSchemaVersion: z.number().int().min(1),
    transformVersion: z.string().min(1),
    rowCount: z.number().int().min(0),
    sourceRuns: z.array(datasetSourceRunSchema).min(1),
    windowDurationSimSeconds: z.number().int().min(1),
    createdAt: z.string().min(1),
    workspaceVersion: z.string().min(1),
    compatibilityMetadata: z.record(z.string(), z.unknown()).optional(),
  })
  .strict()
  .superRefine((value, ctx) => {
    if (value.schemaVersion !== DATASET_MANIFEST_SCHEMA_VERSION) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message: `Unsupported dataset manifest schema version: ${String(value.schemaVersion)}`,
      });
    }
  });

export const featureRejectionSchema = z
  .object({
    eventId: eventIdSchema,
    sequence: sequenceSchema,
    code: featureErrorCodeSchema,
    message: z.string().min(1),
  })
  .strict();

export const onlineFeatureUpdateSchema = z
  .object({
    schemaVersion: z.number().int().min(1),
    windowKey: z.string().min(1),
    vector: featureVectorSchema,
    lastProcessedSequence: sequenceSchema,
    rejections: z.array(featureRejectionSchema).optional(),
  })
  .strict()
  .superRefine((value, ctx) => {
    if (value.schemaVersion !== ONLINE_FEATURE_UPDATE_SCHEMA_VERSION) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message: `Unsupported online feature update schema version: ${String(value.schemaVersion)}`,
      });
    }
  });

export const featureComputeRequestSchema = z
  .object({
    schemaVersion: z.number().int().min(1),
    runId: runIdSchema,
    fromSequence: sequenceSchema.optional(),
    toSequence: sequenceSchema.optional(),
    mode: z.enum(['incremental', 'full']).default('full'),
  })
  .strict();

export const featureComputeResponseSchema = z
  .object({
    schemaVersion: z.number().int().min(1),
    vectors: z.array(featureVectorSchema),
    windows: z.array(featureWindowSchema),
    rejections: z.array(featureRejectionSchema),
    lastProcessedSequence: sequenceSchema,
    outputChecksum: z.string().regex(/^sha256:[0-9a-f]{64}$/),
  })
  .strict();

export const featureParityCheckResponseSchema = z
  .object({
    schemaVersion: z.number().int().min(1),
    matching: z.boolean(),
    offlineChecksum: z.string().regex(/^sha256:[0-9a-f]{64}$/),
    onlineChecksum: z.string().regex(/^sha256:[0-9a-f]{64}$/),
    vectorCount: z.number().int().min(0),
    rejectionCount: z.number().int().min(0),
  })
  .strict();

export type FeatureSchemaManifestV1 = z.infer<typeof featureSchemaManifestSchema>;
export type FeatureVectorV1 = z.infer<typeof featureVectorSchema>;
export type FeatureWindowV1 = z.infer<typeof featureWindowSchema>;
export type FeatureProvenanceV1 = z.infer<typeof featureProvenanceSchema>;
export type DatasetManifestV1 = z.infer<typeof datasetManifestSchema>;
export type OnlineFeatureUpdateV1 = z.infer<typeof onlineFeatureUpdateSchema>;

File: packages/contracts-python/src/aegis_contracts/features.py
"""Phase 14 feature pipeline contracts."""

from __future__ import annotations

from enum import StrEnum
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator

from aegis_contracts.primitives import (
    AuthoredId,
    EventId,
    RunId,
    Sequence,
    SimTimestamp,
    UtcTimestamp,
)
from aegis_contracts.versioning import (
    FEATURE_SCHEMA_MANIFEST_SCHEMA_VERSION,
    assert_supported_schema_version,
)

TRANSFORM_VERSION: str = "0.0.0-phase14"
DEFAULT_WINDOW_DURATION_SIM_SECONDS: int = 300


class FeatureErrorCode(StrEnum):
    VALIDATION_FAILED = "FEATURE_VALIDATION_FAILED"
    UNSUPPORTED_EVENT = "FEATURE_UNSUPPORTED_EVENT"
    DUPLICATE_EVENT = "FEATURE_DUPLICATE_EVENT"
    OUT_OF_ORDER = "FEATURE_OUT_OF_ORDER"
    LATE_EVENT = "FEATURE_LATE_EVENT"
    HIDDEN_TRUTH_BLOCKED = "FEATURE_HIDDEN_TRUTH_BLOCKED"
    SCHEMA_MISMATCH = "FEATURE_SCHEMA_MISMATCH"
    MALFORMED_PAYLOAD = "FEATURE_MALFORMED_PAYLOAD"


class FeatureDataType(StrEnum):
    FLOAT = "float"
    INT = "int"
    CATEGORICAL = "categorical"


class FeatureDefinitionV1(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    name: str = Field(min_length=1, max_length=128)
    order_index: int = Field(alias="orderIndex", ge=0)
    dtype: FeatureDataType
    unit: str = Field(min_length=1, max_length=64)
    min_value: float | None = Field(default=None, alias="minValue")
    max_value: float | None = Field(default=None, alias="maxValue")
    missing_sentinel: float = Field(alias="missingSentinel")
    categorical_mapping_version: int | None = Field(
        default=None, alias="categoricalMappingVersion", ge=1
    )
    categorical_values: list[str] | None = Field(default=None, alias="categoricalValues")


class FeatureSchemaManifestV1(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    schema_version: int = Field(alias="schemaVersion", ge=1)
    feature_schema_version: int = Field(alias="featureSchemaVersion", ge=1)
    transform_version: str = Field(alias="transformVersion", min_length=1)
    window_duration_sim_seconds: int = Field(alias="windowDurationSimSeconds", ge=1)
    features: list[FeatureDefinitionV1] = Field(min_length=1)

    @field_validator("schema_version")
    @classmethod
    def validate_schema_version(cls, value: int) -> int:
        assert_supported_schema_version("feature_schema_manifest", value)
        return value

    @model_validator(mode="after")
    def validate_feature_ordering(self) -> FeatureSchemaManifestV1:
        if self.schema_version != FEATURE_SCHEMA_MANIFEST_SCHEMA_VERSION:
            raise ValueError(
                f"Unsupported feature schema manifest version: {self.schema_version}"
            )
        indices = [feature.order_index for feature in self.features]
        if indices != sorted(indices):
            raise ValueError("Feature definitions must be sorted by orderIndex")
        if len(indices) != len(set(indices)):
            raise ValueError("Feature orderIndex values must be unique")
        return self


class FeatureProvenanceV1(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    run_id: RunId = Field(alias="runId")
    entity_id: AuthoredId = Field(alias="entityId")
    sequence_start: Sequence = Field(alias="sequenceStart", ge=0)
    sequence_end: Sequence = Field(alias="sequenceEnd", ge=0)
    sim_time_start: SimTimestamp = Field(alias="simTimeStart")
    sim_time_end: SimTimestamp = Field(alias="simTimeEnd")
    feature_schema_version: int = Field(alias="featureSchemaVersion", ge=1)
    transform_version: str = Field(alias="transformVersion", min_length=1)
    source_event_ids: list[EventId] = Field(alias="sourceEventIds")


class FeatureWindowV1(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    schema_version: int = Field(alias="schemaVersion", ge=1)
    window_key: str = Field(alias="windowKey", min_length=1)
    run_id: RunId = Field(alias="runId")
    entity_id: AuthoredId = Field(alias="entityId")
    window_start_sim_time: SimTimestamp = Field(alias="windowStartSimTime")
    window_end_sim_time: SimTimestamp = Field(alias="windowEndSimTime")
    is_closed: bool = Field(alias="isClosed")
    watermark_sequence: Sequence = Field(alias="watermarkSequence", ge=0)

    @field_validator("schema_version")
    @classmethod
    def validate_schema_version(cls, value: int) -> int:
        assert_supported_schema_version("feature_window", value)
        return value


class FeatureVectorV1(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    schema_version: int = Field(alias="schemaVersion", ge=1)
    feature_schema_version: int = Field(alias="featureSchemaVersion", ge=1)
    window_key: str = Field(alias="windowKey", min_length=1)
    entity_id: AuthoredId = Field(alias="entityId")
    values: list[float]
    provenance: FeatureProvenanceV1

    @field_validator("schema_version")
    @classmethod
    def validate_schema_version(cls, value: int) -> int:
        assert_supported_schema_version("feature_vector", value)
        return value


class DatasetSourceRunV1(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    run_id: RunId = Field(alias="runId")
    seed: int
    scenario_version_id: AuthoredId = Field(alias="scenarioVersionId")


class DatasetManifestV1(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    schema_version: int = Field(alias="schemaVersion", ge=1)
    dataset_checksum: str = Field(alias="datasetChecksum", pattern=r"^sha256:[0-9a-f]{64}$")
    feature_schema_version: int = Field(alias="featureSchemaVersion", ge=1)
    transform_version: str = Field(alias="transformVersion", min_length=1)
    row_count: int = Field(alias="rowCount", ge=0)
    source_runs: list[DatasetSourceRunV1] = Field(alias="sourceRuns", min_length=1)
    window_duration_sim_seconds: int = Field(alias="windowDurationSimSeconds", ge=1)
    created_at: UtcTimestamp = Field(alias="createdAt")
    workspace_version: str = Field(alias="workspaceVersion", min_length=1)
    compatibility_metadata: dict[str, Any] = Field(
        default_factory=dict, alias="compatibilityMetadata"
    )

    @field_validator("schema_version")
    @classmethod
    def validate_schema_version(cls, value: int) -> int:
        assert_supported_schema_version("dataset_manifest", value)
        return value


class FeatureRejectionV1(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    event_id: EventId = Field(alias="eventId")
    sequence: Sequence
    code: FeatureErrorCode
    message: str = Field(min_length=1)


class OnlineFeatureUpdateV1(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    schema_version: int = Field(alias="schemaVersion", ge=1)
    window_key: str = Field(alias="windowKey", min_length=1)
    vector: FeatureVectorV1
    last_processed_sequence: Sequence = Field(alias="lastProcessedSequence", ge=0)
    rejections: list[FeatureRejectionV1] = Field(default_factory=list)

    @field_validator("schema_version")
    @classmethod
    def validate_schema_version(cls, value: int) -> int:
        assert_supported_schema_version("online_feature_update", value)
        return value


class FeatureComputeRequestV1(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    schema_version: int = Field(alias="schemaVersion", ge=1)
    run_id: RunId = Field(alias="runId")
    from_sequence: Sequence | None = Field(default=None, alias="fromSequence", ge=0)
    to_sequence: Sequence | None = Field(default=None, alias="toSequence", ge=0)
    mode: Literal["incremental", "full"] = "full"


class FeatureComputeResponseV1(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    schema_version: int = Field(alias="schemaVersion", ge=1)
    vectors: list[FeatureVectorV1]
    windows: list[FeatureWindowV1]
    rejections: list[FeatureRejectionV1]
    last_processed_sequence: Sequence = Field(alias="lastProcessedSequence", ge=0)
    output_checksum: str = Field(alias="outputChecksum", pattern=r"^sha256:[0-9a-f]{64}$")


class FeatureParityCheckResponseV1(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    schema_version: int = Field(alias="schemaVersion", ge=1)
    matching: bool
    offline_checksum: str = Field(alias="offlineChecksum", pattern=r"^sha256:[0-9a-f]{64}$")
    online_checksum: str = Field(alias="onlineChecksum", pattern=r"^sha256:[0-9a-f]{64}$")
    vector_count: int = Field(alias="vectorCount", ge=0)
    rejection_count: int = Field(alias="rejectionCount", ge=0)
